src/main/cpp/src/get_json_object.cu

```diff
@@ -789,207 +790,391 @@ __device__ thrust::pair<bool, cudf::size_type> evaluate_path(
-        [in_offsets] __device__(auto const idx) { return in_offsets[idx + 1] - in_offsets[idx]; }),
-      int64_t{0},
-      thrust::maximum{});
-
+  auto const scratch_size = [&, max_row_size = max_row_size] {
     // Pad the scratch buffer by an additional size that is a multiple of max row size.
     auto constexpr padding_rows = 10;
     return input.chars_size(stream) + max_row_size * padding_rows;
   }();
-  auto output_scratch  = rmm::device_uvector<char>(scratch_size, stream);
-  auto out_stringviews = rmm::device_uvector<thrust::pair<char const*, cudf::size_type>>{
-    static_cast<std::size_t>(input.size()), stream};
-  auto has_out_of_bound = rmm::device_scalar<bool>{false, stream};
-
-  constexpr int blocks_per_SM = 1;
-  constexpr int block_size    = 256;
-  auto const num_blocks       = [&] {
-    int device_id{};
-    cudaDeviceProp props{};
-    CUDF_CUDA_TRY(cudaGetDevice(&device_id));
-    CUDF_CUDA_TRY(cudaGetDeviceProperties(&props, device_id));
-    return props.multiProcessorCount * blocks_per_SM;
-  }();
 
-  get_json_object_kernel<block_size>
-    <<<num_blocks, block_size, 0, stream.value()>>>(*d_input_ptr,
-                                                    in_offsets,
-                                                    d_path_commands,
-                                                    out_stringviews.data(),
-                                                    output_scratch.data(),
-                                                    has_out_of_bound.data());
+  rmm::device_uvector<int8_t> d_has_out_of_bound(num_outputs, stream);
+  std::vector<rmm::device_uvector<char>> scratch_buffers;
+  std::vector<rmm::device_uvector<thrust::pair<char const*, cudf::size_type>>> out_stringviews;
+  std::vector<json_path_processing_data> h_path_data;
+  scratch_buffers.reserve(json_paths.size());
+  out_stringviews.reserve(json_paths.size());
+  h_path_data.reserve(json_paths.size());
+
+  for (std::size_t idx = 0; idx < num_outputs; ++idx) {
+    auto const& instructions = json_paths[idx];
+    if (instructions.size() > max_path_depth) { CUDF_FAIL("JSONPath query exceeds maximum depth"); }
+
+    scratch_buffers.emplace_back(rmm::device_uvector<char>(scratch_size, stream));
+    out_stringviews.emplace_back(rmm::device_uvector<thrust::pair<char const*, cudf::size_type>>{
+      static_cast<std::size_t>(input.size()), stream});
+
+    h_path_data.emplace_back(json_path_processing_data{d_json_paths[idx],
+                                                       in_offsets,
+                                                       out_stringviews.back().data(),
+                                                       scratch_buffers.back().data(),
+                                                       d_has_out_of_bound.data() + idx});
+  }
+  auto d_path_data = cudf::detail::make_device_uvector_async(
+    h_path_data, stream, rmm::mr::get_current_device_resource());
+  thrust::uninitialized_fill(
```

Comment:
Running on GPU, because of the number of json paths is not too large, running CPU will be more fast.